pkg: ndn-riot: initial import of ndn-riot as a package

cgundogan opened the pull request:
This PR imports `ndn-riot` [1] as a `pkg`.
Furthermore, I included a tests application that I adopted from [2].

@cawka I hope you don't mind that I included your changes proposed in #5648.

Can be tested by navigating into `tests/ndn-ping` and invoke the usual `make all term PORT=tap0/1`

[1] https://github.com/named-data-iot/ndn-riot
[2] https://github.com/named-data-iot/ndn-riot-examples/tree/master/ndn-ping

mfrey commented:
Build fails for me with (I'm a bit puzzled since the next statement is a assert using the variable)
```
/home/frey/Desktop/Projekte/RIOT/tests/ndn-ping/bin/pkg/native/ndn-riot/pit.c: In function ‘ndn_pit_add’:
/home/frey/Desktop/Projekte/RIOT/tests/ndn-ping/bin/pkg/native/ndn-riot/pit.c:104:13: error: unused variable ‘r’ [-Werror=unused-variable]
         int r = ndn_interest_get_name(&entry->shared_pi->block, &pn);
```

OlegHahm commented:
Looking forward to test this.

OlegHahm commented:
@mfrey, maybe you're building (the package) with `NDEBUG` set?

mfrey commented:
@OlegHahm I haven't set anything actually. The variable is only used in the assert statement. If you don't build with NEDBUG this is an issue nontheless. Isn't it? Anyway, @cgundogan fixed it by a cast to void. I'll give it a try later.

OlegHahm commented:
> The variable is only used in the assert statement. If you don't build with NEDBUG this is an issue nontheless. Isn't it?

Sure. Just wanted to find an explanation.

wentaoshang commented:
There is also a unittest suite for encoding & decoding APIs that has not been merged into ndn-riot repo yet:

https://github.com/wentaoshang/RIOT/tree/ndn/tests/unittests/tests-ndn_encoding

I can try to make a PR for this if you think it can be included.

cgundogan commented:
>if you think it can be included.

nice, definitely!

OlegHahm commented:
Doesn't ndn-riot require malloc/free? In that case, the application should use the tlsf package.

OlegHahm commented:
Until #5943 is fixed, the applications should also blacklist boards without hwrng.

OlegHahm commented:
@cgundogan, with https://github.com/named-data-iot/ndn-riot/issues/2 being merged, could you please update the revision of the package?

cgundogan commented:
@OlegHahm sure, sorry for the delay. I amended the new commit hash to point to current master of https://github.com/named-data-iot/

wentaoshang commented:
Sorry I lost track of this issue. Wonder what is the current status? Anything needed to be fixed from NDN-RIOT side? Honestly I'm not very familiar with the Github collaboration procedure but I can try my best to help.

Also I have some concern about the requirement for hardware RNG for micro-ecc, because the two platforms (SAMR21 and IoT-Lab M3) I'm using for NDN-RIOT right now do not have this feature. What I did is to patch micro-ecc (https://github.com/named-data-iot/RIOT/commit/8b9f0b1f40d56b887dbe54651c849ccb511bf4f3) and use deterministic ECDSA signing with pre-generated keys (RFC6979). Any comment on this idea?

cgundogan commented:
@OlegHahm ping?

@wentaoshang I have no experience with the micro-ecc package, but the diff in your commit (https://github.com/named-data-iot/RIOT/commit/8b9f0b1f40d56b887dbe54651c849ccb511bf4f3) is not very apparent to me. What exactly did you change?

tfar commented:
@wentaoshang so you're using your own deterministic ECDSA implementation or is the one you're using part of MicroECC? The patch you're referencing, named-data-iot@8b9f0b1 , removes the RIOT hardware random generator integration. Does MicroECC behave differently based on whether `g_rng_function` is defined or not; or why was it removed? I think RIOT's [hwrng_read](https://riot-os.org/api/group__drivers__periph__hwrng.html#ga860fd64c46fd05d1e7adc2222da5d060) is always available, regardless of hardware support.

wentaoshang commented:
Sorry I forgot to mention this: I'm using a signing function called `uECC_sign_deterministic` in MicroECC https://github.com/kmackay/micro-ecc/blob/master/uECC.c#L1377 . For that function to work, `g_rng_function` has to be unset (hence the removal of old patch 0002). The side effect is that it'll turn off hardware RNG for all platforms, which is bad. I tried to figure out some preprocessor tricks to fix that but failed...

tfar commented:
@wentaoshang So instead of removing the default implementation, can't you call `void uECC_set_rng(uECC_RNG_Function rng_function)` at https://github.com/kmackay/micro-ecc/blob/master/uECC.c#L190 in your user code to reset the random number generator callback?

wentaoshang commented:
That could also work... Ideally I want to be able to switch automatically between deterministic/nondeterministic signing (maybe at compile time?), based on whether HWRNG is available. But what you proposed sounds a good solution for now.

tfar commented:
The setter/getter functions for the random generator callback are part of the public API https://github.com/kmackay/micro-ecc/blob/master/uECC.h#L127 . So it seems what you want to achieve is possible without completely removing support for hardware random number generator in MciroECC.

wentaoshang commented:
I agree. I'll try to figure out how to update ndn-riot code. I remember that without named-data-iot@8b9f0b1 the code could not compile for platforms with no HWRNG because `hwrng_read` was not defined. Not sure if the problem has been fixed in more recent releases.

wentaoshang commented:
Hi all, given that #6204 has been merged, I wonder how we can move this PR forward? I'd be more than happy to help.

OlegHahm commented:
> OlegHahm added the Hack'n'ACK candidate label 7 days ago

I was planning to work on it, but obviously I wasn't very successful. ;-) Still on my list though.

wentaoshang commented:
If you like, I can try to update our old PR #5648 to make it consistent with this one, then we can start from there.

cgundogan commented:
@wentaoshang @OlegHahm sorry for the delay! I updated the package version to point to current master of ndn-riot https://github.com/named-data-iot/ndn-riot/commit/fd165856fd19b8e0d6b435ecc82b485727ef842e. The `ndnping` test application still seems to work (tested with native).

wentaoshang commented:
Thanks! A small comment: I would put ndn-ping under the examples folder rather than tests folder, given that it is an important utility in NDN.

cgundogan commented:
>given that it is an important utility in NDN

ping is an important utility in NDN? (:

I have no strong opinion whether it resides in the `examples` folder or the `tests` folder. The ping application could indeed also be a showcase for the ndn support in RIOT. I will move the application later today or tomorrow into `examples` 👍

wentaoshang commented:
Yeah... kind of surprising but ping is one of the earliest apps ever implemented over NDN (CCN at that time ;).

Spoiler alert: there is on-going design effort for NDN traceroute, too. Maybe you've heard about that at ICNRG. I'll try to port that to RIOT in the future, once the design is finalized.

cgundogan commented:
>Spoiler alert: there is on-going design effort for NDN traceroute, too. Maybe you've heard about that at ICNRG. I'll try to port that to RIOT in the future, once the design is finalized.

Yeah I read about that on ICNRG's mailing list (: Would be nice to see that in RIOT :+1:

cgundogan commented:
@wentaoshang I moved the application to `examples`

wentaoshang commented:
It looks good to me 👍  Looking forward to this PR being merged :)

smlng commented on examples/ndn-ping/ndn_ping.c:
```diff
@@ -0,0 +1,327 @@
+           *(content.buf + 4), *(content.buf + 5));
+
+    r = ndn_data_verify_signature(data, ecc_key_pub, sizeof(ecc_key_pub));
+    if (r != 0) {
```
shouldn't this return an error code, e.g. `NDN_APP_ERROR`

wentaoshang replied:
Depending on the application logic, sometimes you may still want to process packets with invalid signatures (e.g., signed by an expired but valid key). In this particular case, however, it makes sense to me to stop immediately and return `NDN_APP_ERROR`.

cgundogan replied:
IMO these changes should be addressed in follow-up PRs. My current PR is just a simple import from the original sources.

smlng commented on examples/ndn-ping/ndn_ping.c:
```diff
@@ -0,0 +1,327 @@
+{
+    ndn_block_t name;
+    ndn_interest_get_name(interest, &name);
+    assert(r == 0);
```
which `r` is asserted here?

wentaoshang replied:
Should be the return value from `ndn_interest_get_name`: https://github.com/named-data-iot/ndn-riot-examples/blob/master/ndn-ping/ndn_ping.c#L91

OlegHahm replied:
But the `int r = ...` in the line above is missing. ;)

cgundogan replied:
Ok, @wentaoshang I imported the newest version of `ndn-ping` from https://github.com/named-data-iot/ndn-riot-examples/tree/master/ndn-ping and adapted the files to adhere our coding conventions (no tabs, indentations, C style comments, etc.)

wentaoshang commented:
I updated the ndn-riot code to use C11 atomic, to be consistent with the latest RIOT-OS master branch. Could you update the `PKG_VERSION` in this patch to point to the latest commit of ndn-riot?

OlegHahm commented on examples/ndn-ping/ndn_ping.c:
```diff
@@ -0,0 +1,326 @@
+
+static ndn_app_t* handle = NULL;
+
+/* static const unsigned char key[] = { 'd', 'u', 'm', 'm', 'y', 'k', 'e', 'y' }; */
```
What's the rationale behind this?

wentaoshang replied:
That dummy key is used for implementing HMAC signature (used by some earlier version of this code). It could be removed.

OlegHahm commented on pkg/ndn-riot/Makefile:
```diff
@@ -0,0 +1,9 @@
+PKG_NAME=ndn-riot
+PKG_URL=https://github.com/named-data-iot/ndn-riot
+PKG_VERSION=fd165856fd19b8e0d6b435ecc82b485727ef842e
+
```
Can you add `PKG_LICENSE` please?

OlegHahm commented:
The example lacks a README.

OlegHahm commented:
```
In file included from /home/oleg/git/RIOT/examples/ndn-ping/bin/pkg/samr21-xpro/ndn-riot/encoding/name.h:23:0,
                 from /home/oleg/git/RIOT/examples/ndn-ping/bin/pkg/samr21-xpro/ndn-riot/app.h:23,
                 from /home/oleg/git/RIOT/examples/ndn-ping/ndn_ping.c:27:
/home/oleg/git/RIOT/examples/ndn-ping/bin/pkg/samr21-xpro/ndn-riot/encoding/shared-block.h:25:20: fatal error: atomic.h: No such file or directory
 #include <atomic.h>
```

wentaoshang commented:
The issue with <atomic.h> is already fixed: [see this commit](https://github.com/named-data-iot/ndn-riot/commit/4a2c0774ab2f8f42acc3340e405487661991422b)

aabadie commented:
@cgundogan @OlegHahm what's the status here. Any chance for you to make it for the release ?

wentaoshang commented:
The latest ndn-riot code makes use of the CCM encryption mode. Please also merge #6706.

astralien3000 commented:
Hello ! I will try to push forward this PR forward.
I tested to compile ndn-ping example for native, which did not work because of `atomic.h`, which I have heard of in this thread. Is the branch up to date ?
I tried for samr21-xpro, and it compiles (but not tested on real hardware).

Just for reference, I was able to make ndn-riot work with RIOT in this [RIOT version](https://github.com/astralien3000/RIOT/commit/c366d7b8e15a6c923570fbedeccf226f89cfe929) and this [ndn-riot version](https://github.com/astralien3000/ndn-riot/tree/532eb723869a11ca4f95f6fc2ecd734fa4567069), on native and samr21-xpro (compiles and runs correctly).

I see that this PR is "dependent" of other PRs ( #6706 and #5648 ) that are not merged nor closed. Is this branch including those changes ?

miri64 commented:
@cgundogan ping, needs rebase ;-)

astralien3000 commented:
@miri64, @cgundogan, this may need more than a rebase due to API change (see #8415).
I had no time to work on this, sorry.

astralien3000 commented:
I made a [PR to the ndn-riot repos](https://github.com/named-data-iot/ndn-riot/pull/9) that should solve the problem(s).
If you want to test it, you can download the package with : 
```Makefile
PKG_URL=https://github.com/astralien3000/ndn-riot
PKG_VERSION=4d258a283d50cd34fc9074479a9fa32e0dce57ff
```

cgundogan commented:
@astralien3000 I rebased to current master and updated the ndn-riot commit hash to point to their current master (including your fixes). Could you give this a shot?

astralien3000 commented:
Sorry for the late answer,

I tested your branch, the example `ndnping` works well.
I guess we could consider merging this now ? ACK from my side.

mfrey commented:
> I guess we could consider merging this now ? ACK from my side.

As far as I can see it still lacks a README (as @OlegHahm pointed out). Also, there seems to be a a conflict with ``sys/auto_init/auto_init.c``. 

Let's hope that we don't have to celebrate 🍰 for this PR "open for 2 years".

aabadie commented on Makefile.dep:
```diff
@@ -14,6 +14,14 @@ ifneq (,$(filter cbor_ctime,$(USEMODULE)))
 endif
 
+ifneq (,$(filter ndn-riot,$(USEPKG)))
+    USEMODULE += gnrc
```
My 2c style comment: indentation should be 2 spaces, not 4

aabadie commented on examples/ndn-ping/README.md:
```diff
@@ -0,0 +1,49 @@
+# ndn-ping
+
+This application demonstrates the usage of the package ndn-riot.
+This example basically enable the user to setup a ndn data server, and a ndn client that can request the data.
```
enable**s**

aabadie commented on examples/ndn-ping/README.md:
```diff
@@ -0,0 +1,49 @@
+
+# Setting up for native
+
+You need to create `tap` and `tapbr` devices using RIOT's `tapsetup` script before stating the application:
```
Below an affirmative style is used, I would do the same here as well: `Create tap and tapbr, etc`

aabadie commented on examples/ndn-ping/README.md:
````diff
@@ -0,0 +1,49 @@
+./RIOTDIR/dist/tools/tapsetup/tapsetup
+```
+
+Then you can run the application on 2 different terminals :
````
Same here: `Then run the application...`

aabadie commented on examples/ndn-ping/main.c:
```diff
@@ -0,0 +1,43 @@
+ * @file
+ * @brief       NDN ping application
+ *
+ * @author      Wentao Shang <wentaoshang@gmaiil.com>
```
s/gmaiil/gmail/ ?

There might be others like this

aabadie commented on examples/ndn-ping/ndn_ping.c:
```diff
@@ -0,0 +1,326 @@
+           *(content.buf + 4), *(content.buf + 5));
+
+    r = ndn_data_verify_signature(data, ecc_key_pub, sizeof(ecc_key_pub));
+    if (r != 0)
```
This is not inline with the coding conventions; missing opening and closing curly braces and indentation is off on the line below

astralien3000 commented:
Comments addressed (thanks @aabadie for the comments), and squashed.

aabadie commented on examples/ndn-ping/ndn_ping.c:
```diff
@@ -0,0 +1,329 @@
+
+/* static const unsigned char key[] = { 'd', 'u', 'm', 'm', 'y', 'k', 'e', 'y' }; */
+
+static const uint8_t ecc_key_pri[] = {
```
Are those keys always the same or should they be set by the user ? If it's the second case, then you should find a way to make them configurable (at build time or by using a setter function)

astralien3000 replied:
Well, the user of an android don't choose the keys of his apps.
And if you want build-time configuration, I would do it the current way.

cgundogan commented:
@OlegHahm is your change request still valid?

astralien3000 commented:
@cgundogan With my last commit, it should not be. Is it right @OlegHahm  ?

astralien3000 commented:
Well, I don't think we will have any news. Is everyone ok to dismiss the review of @OlegHahm ? Or do we wait 3 months more to miss the next release ?

mfrey commented:
> Well, I don't think we will have any news. Is everyone ok to dismiss the review of @OlegHahm ? 

I'm in favour of dismissing @OlegHahm review. Any thoughts  @aabadie @wentaoshang @smlng?

aabadie commented:
> I'm in favour of dismissing @OlegHahm review. 

I sent him a private mail yesterday. Let's way a few days until next week. If he doesn't reply, then I'm also in favor of dismissing his review.

OlegHahm commented:
Not sure how to do this via mobile but I don't have any further requests if the other reviewers are fine. And I'm very sorry for the delay that was caused by my unresponsiveness.

aabadie commented:
I activated Murdock. @cgundogan @astralien3000  do you ACK ?

astralien3000 commented:
I can't reproduce the error of Murdock, which is weird (it's not a compilation error, but a directory that is missing). Any idea to help me ? @kaspar030 ?

gebart commented:
It looks like a race condition. Try locally with make -j9

kaspar030 commented:
> I can't reproduce the error of Murdock, which is weird (it's not a compilation error, but a directory that is missing). Any idea to help me ? @kaspar030 ?

Try to do build the package in "pkg/ndn-riot/Makefile" (instead of adding "DIRS += BINDIRBASE/..." in "pkg/ndn-riot/Makefile.include". See how other packages do it, usually sth like ```"$(MAKE)" -C $(PKG_BUILDDIR) ```.

astralien3000 commented:
Thank you @kaspar030, the problem was indeed the missing rule for compiling the package. I still don't understand why it works on my computer an not for Murdock, but well, at least it is fixed.

cgundogan commented on pkg/ndn-riot/Makefile:
```diff
@@ -0,0 +1,11 @@
+PKG_NAME=ndn-riot
+PKG_URL=https://github.com/astralien3000/ndn-riot
```
What is the reason not to take the official repo's url?

astralien3000 replied:
New errors appeared since the last time I tried the code. I will need to do a pull request to the official repos, but for the moment, I try with mine. Of course this should not be merged for the moment.

cawka replied:
Just in case, I can quickly merge pull requests in named-data-iot/ndn-riot repo

astralien3000 commented:
@kaspar030, well actually adding what you said did not help. I don't know why it worked once, but I got the same error again for 3 builds of Murdock... Maybe it's a race condition like @gebart said ? But I can't reproduce the error even with -j9...

kaspar030 commented on pkg/ndn-riot/Makefile.include:
```diff
@@ -0,0 +1,3 @@
+INCLUDES += -I$(BINDIRBASE)/pkg/$(BOARD)
+DIRS += $(BINDIRBASE)/pkg/$(BOARD)/ndn-riot
```
this line needs to go

astralien3000 commented:
Ok ! Almost there ! I just need to wait for [the PR of ndn-riot](https://github.com/named-data-iot/ndn-riot/pull/13), and we can move. Maybe tomorrow.

astralien3000 commented:
@cgundogan, @mfrey, can one of you test ndn-ping ?

aabadie commented on sys/include/net/ethertype.h:
```diff
@@ -34,7 +34,8 @@ extern "C" {
 #define ETHERTYPE_RESERVED      (0x0000)    /**< Reserved */
 #define ETHERTYPE_IPV4          (0x0800)    /**< Internet protocol version 4 */
 #define ETHERTYPE_ARP           (0x0806)    /**< Address resolution protocol */
-#define ETHERTYPE_NDN           (0x0801)    /**< Parc CCNX */
```
Could this change have an impact on other named data network (CCN-lite) implementations ?

mfrey replied:
I don't think so. CCN-lite for RIOT uses only NDN for now

cawka replied:
Actually, this should be changed to 
```
#define ETHERTYPE_NDN 0x8624
```

This is the value we are using in other places.  This, probably, was just a quick hack.

mfrey replied:
I'm slightly confused by the PARC CCNX comment.

astralien3000 replied:
I just tested the ccn-lite-relay example, it behave the same as on master.

astralien3000 replied:
actually this line is the one deleted, ETHERTYPE_NDN is defined as 0x8624 in this PR.
but ETHERTYPE_CCNX is still defined as 0x0801. What do we do about these lines ?
